### PR TITLE
RNN high level API

### PR DESCRIPTION
**Pull request by unnonouno**
I extracted high-level API from Chainer. It provides high-level API for cudnn's RNN.

### CODE REVIEW

File: cupy/cudnn.pyx [hunk 1]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+
+    def append(self, desc):
+        self._value.push_back(desc)
+
+    @property
+    def data(self):
+        return <size_t>&self._value[0]
+    
+
+cpdef DescriptorArray _make_tensor_descriptor_array(xs, lengths):
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
cdef DescriptorArray _make_tensor_descriptor_array(xs, lengths):
```

File: cupy/cudnn.pyx [hunk 2]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+
+    return descs
+
+
+cdef core.ndarray _make_rnn_workspace(
+    Descriptor rnn_desc, int length, DescriptorArray descs):
+    cdef size_t handle = get_handle()
+    cdef size_t work_size = py_cudnn.getRNNWorkspaceSize(
+        handle, rnn_desc.value, length, descs.data)
+    return core.ndarray((work_size,), dtype='b')
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
    return memory.alloc(work_size)
```

File: cupy/cudnn.pyx [hunk 3]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+        cx_ptr = 0
+        cx_desc_value = 0
+        cy = None
+        cy_ptr = 0
+        cy_desc_value = 0
+
+    cdef core.ndarray workspace = _make_rnn_workspace(
+        rnn_desc, length, c_x_descs)
+    
+    py_cudnn.RNNForwardInference(
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
    cudnn.RNNForwardInference(
```

File: cupy/cudnn.pyx [hunk 4]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+        cy = None
+        cy_ptr = 0
+        cy_desc_value = 0
+
+    cdef core.ndarray workspace = _make_rnn_workspace(
+        rnn_desc, length, xs_descs)
+    
+    cdef size_t reserve_size = py_cudnn.getRNNTrainingReserveSize(
+        handle, rnn_desc.value, length, xs_descs.data)
+    cdef core.ndarray reserve_space = core.ndarray((reserve_size,), dtype='b')
```
Comment (okuta) [inline cupy/cudnn.pyx]: What is `reserve_space`?

File: cupy/cudnn.pyx [hunk 5]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+    ys = core.ascontiguousarray(ys)
+    dhy = core.ascontiguousarray(dhy)
+    dcy = core.ascontiguousarray(dcy)
+    dys = core.ascontiguousarray(dys)
+
+    cdef int length = len(lengths)
+    cdef int n_layers
+    cdef int n_units = hx.shape[2]
+
+    if direction_mode == py_cudnn.CUDNN_BIDIRECTIONAL:
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
    if direction_mode == cudnn.CUDNN_BIDIRECTIONAL:
```

File: cupy/cudnn.pyx [hunk 6]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+    hx = core.ascontiguousarray(hx)
+    cx = core.ascontiguousarray(cx)
+    w = core.ascontiguousarray(w)
+    xs = core.ascontiguousarray(xs)
+
+    cdef int length = len(lengths)
+    cdef int n_layers
+    cdef int n_units = hx.shape[2]
+    cdef int input_units
+    if direction_mode == py_cudnn.CUDNN_BIDIRECTIONAL:
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
    if direction_mode == cudnn.CUDNN_BIDIRECTIONAL:
```

File: cupy/cudnn.pyx [hunk 7]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+        cudnn.setTensorNdDescriptor(
+            desc, data_type, 3, <size_t>&c_shape[0], <size_t>&c_strides[0])
+
+    return descs
+
+
+cdef core.ndarray _make_rnn_workspace(
+    Descriptor rnn_desc, int length, DescriptorArray descs):
+    cdef size_t handle = get_handle()
+    cdef size_t work_size = py_cudnn.getRNNWorkspaceSize(
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
    cdef size_t work_size = cudnn.getRNNWorkspaceSize(
```

File: cupy/cudnn.pyx [hunk 8]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+    else:  # py_cudnn.CUDNN_UNIDIRECTIONAL
+        input_units = n_units
+        n_layers = hx.shape[0]
+
+    cdef core.ndarray ys = core.ndarray((len(xs), input_units), dtype=xs.dtype)
+    cdef size_t handle = get_handle()
+
+    cdef Descriptor rnn_desc = create_rnn_descriptor(
+        n_units, n_layers, states._desc,
+        py_cudnn.CUDNN_LINEAR_INPUT, direction_mode,
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
        cudnn.CUDNN_LINEAR_INPUT, direction_mode,
```

File: cupy/cudnn.pyx [hunk 9]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+    cx = core.ascontiguousarray(cx)
+    w = core.ascontiguousarray(w)
+    xs = core.ascontiguousarray(xs)
+
+    cdef int length = len(lengths)
+    cdef int n_layers
+    cdef int n_units = hx.shape[2]
+    cdef int input_units
+
+    if direction_mode == py_cudnn.CUDNN_BIDIRECTIONAL:
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
    if direction_mode == cudnn.CUDNN_BIDIRECTIONAL:
```

File: cupy/cudnn.pyx [hunk 10]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+        input_units = n_units * 2
+        n_layers = hx.shape[0] // 2
+    else:  # py_cudnn.CUDNN_UNIDIRECTIONAL
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
    else:  # cudnn.CUDNN_UNIDIRECTIONAL
```

File: cupy/cudnn.pyx [hunk 11]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+        input_units = n_units
+        n_layers = hx.shape[0]
+
+    cdef core.ndarray ys = core.ndarray((len(xs), input_units), dtype=xs.dtype)
+    cdef size_t handle = get_handle()
+
+    cdef Descriptor rnn_desc = create_rnn_descriptor(
+        n_units, n_layers, states._desc,
+        py_cudnn.CUDNN_LINEAR_INPUT, direction_mode,
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
        cudnn.CUDNN_LINEAR_INPUT, direction_mode,
```

File: cupy/cudnn.pyx [hunk 12]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+
+    if direction_mode == py_cudnn.CUDNN_BIDIRECTIONAL:
+        n_layers = hx.shape[0] // 2
+    else:  # py_cudnn.CUDNN_UNIDIRECTIONAL
+        n_layers = hx.shape[0]
+
+    cdef size_t handle = get_handle()
+    cdef Descriptor rnn_desc = create_rnn_descriptor(
+        n_units, n_layers, states._desc,
+        py_cudnn.CUDNN_LINEAR_INPUT, direction_mode,
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
        cudnn.CUDNN_LINEAR_INPUT, direction_mode,
```

File: cupy/cudnn.pyx [hunk 13]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+    else:
+        cx_ptr = 0
+        cx_desc_value = 0
+        dcx = None
+        dcx_ptr = 0
+        dcx_desc_value = 0
+        dcy_ptr = 0
+        dcy_desc_value = 0
+
+    py_cudnn.RNNBackwardData(
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
    cudnn.RNNBackwardData(
```

File: cupy/cudnn.pyx [hunk 14]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+    xs = core.ascontiguousarray(xs)
+    hx = core.ascontiguousarray(hx)
+    ys = core.ascontiguousarray(ys)
+    w = core.ascontiguousarray(w)
+
+    cdef int length = len(lengths)
+    cdef int n_layers
+    cdef int n_units = hx.shape[2]
+
+    if direction_mode == py_cudnn.CUDNN_BIDIRECTIONAL:
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
    if direction_mode == cudnn.CUDNN_BIDIRECTIONAL:
```

File: cupy/cudnn.pyx [hunk 15]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+        n_layers = hx.shape[0] // 2
+    else:  # py_cudnn.CUDNN_UNIDIRECTIONAL
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
    else:  # cudnn.CUDNN_UNIDIRECTIONAL
```

File: cupy/cudnn.pyx [hunk 16]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+        n_layers = hx.shape[0]
+
+    cdef size_t handle = get_handle()
+    cdef Descriptor rnn_desc = create_rnn_descriptor(
+        n_units, n_layers, states._desc,
+        py_cudnn.CUDNN_LINEAR_INPUT, direction_mode,
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
        cudnn.CUDNN_LINEAR_INPUT, direction_mode,
```

File: cupy/cudnn.pyx [hunk 17]
```diff
@@ -437,6 +440,328 @@ def get_rnn_lin_layer_bias_params(
...
+    cdef Descriptor hx_desc = create_tensor_nd_descriptor(hx)
+
+    cdef core.ndarray workspace = _make_rnn_workspace(
+        rnn_desc, length, xs_descs)
+
+    cdef core.ndarray dw = core.ndarray(w.shape, w.dtype)
+    dw[...] = 0
+    cdef Descriptor dw_desc = create_filter_descriptor(dw)
+
+    py_cudnn.RNNBackwardWeights(
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
    cudnn.RNNBackwardWeights(
```

File: cupy/cudnn.pyx [hunk 18]
```diff
@@ -573,7 +898,11 @@ cpdef bint is_tensor_core_available(dtype) except *:
             int(device.get_compute_capability()) == 70)
 
 
-class DropoutStates(object):
+cdef class DropoutStates(object):
```
Comment (okuta) [inline cupy/cudnn.pyx]:
```suggestion
cdef class DropoutStates:
```

**Comment by unnonouno** [comment]
I want to discuss some API design:
- Do we need to call `ascontiguousarray` in APIs? (do we need to support non-contiguous arrays?)
- Is it OK to expose enum values such as `cudnn.CUDNN_BIDIRECTIONAL`?

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**